rascal2/core/runner.py

```diff
@@ -0,0 +1,108 @@
+"""QObject for running RAT."""
+
+from dataclasses import dataclass
+from multiprocessing import Process, Queue
+
+import RATapi as RAT
+from PyQt6 import QtCore
+from RATapi.utils.enums import Procedures
+
+
+class RATRunner(QtCore.QObject):
+    """Class for running RAT."""
+
+    event_received = QtCore.pyqtSignal()
+    finished = QtCore.pyqtSignal()
+    stopped = QtCore.pyqtSignal()
+
+    def __init__(self, rat_inputs, procedure: Procedures, display_on: bool):
+        super().__init__()
+        self.timer = QtCore.QTimer()
+        self.timer.setInterval(1)
+        self.timer.timeout.connect(self.check_queue)
+
+        # this queue handles both event data and results
+        self.queue = Queue()
+
+        self.process = Process(target=run, args=(self.queue, rat_inputs, procedure, display_on))
+
+        self.results = None
+        self.error = None
+        self.events = []
+
+    def start(self):
+        """Start the calculation."""
+        self.process.start()
+        self.timer.start()
+
+    def interrupt(self):
+        """Interrupt the running process."""
+        self.timer.stop()
+        self.process.kill()
+        self.stopped.emit()
+
+    def check_queue(self):
+        """Check for new data in the queue."""
+        if not self.process.is_alive():
+            self.timer.stop()
+        self.queue.put(None)
+        # the queue should never be empty at this line (because we just put a
+        # None in it) but if it is, it'd hang forever - we add the timeout
+        # to avoid this happening just in case
+        for item in iter(lambda: self.queue.get(timeout=500), None):
```

Comment:
I agree we never want the software to hang forever but I can't think of a case when this timeout would ever be useful. Did you have something specific in mind when adding the timeout? I am fine to leave it in just curious

Reply:
at some point while writing this code I was encountering a weird race condition with the queue. that seems to have gone away now so i'm going to remove the timeout so that any future code which adds a race condition will be noticeably broken